xfields/beam_elements/beambeam_src/beambeam3d.h

```diff
@@ -155,35 +200,87 @@ void BeamBeamBiGaussian3D_track_local_particle(BeamBeamBiGaussian3DData el,
     const double post_subtract_zeta = BeamBeamBiGaussian3DData_get_post_subtract_zeta(el);
     const double post_subtract_pzeta = BeamBeamBiGaussian3DData_get_post_subtract_pzeta(el);
 
+    // Extract the record and record_index
+    BeamBeamBiGaussian3DRecordData record = BeamBeamBiGaussian3DData_getp_internal_record(el, part0);
+    BeamstrahlungTableData table = NULL;
+    RecordIndex table_index = NULL;
+    if (record){
+        table = BeamBeamBiGaussian3DRecordData_getp_beamstrahlungtable(record);
+        table_index = BeamstrahlungTableData_getp__index(table);
+    }
+
     //start_per_particle_block (part0->part)
         double x = LocalParticle_get_x(part);
         double px = LocalParticle_get_px(part);
         double y = LocalParticle_get_y(part);
         double py = LocalParticle_get_py(part);
         double zeta = LocalParticle_get_zeta(part);
         double pzeta = LocalParticle_get_pzeta(part);
+        //double delta = LocalParticle_get_delta(part);
 
         const double q0 = LocalParticle_get_q0(part);
         const double p0c = LocalParticle_get_p0c(part); // eV
 
-
+/*
+        printf("[beambeam3d] [%d] before boost:\n", part->ipart);
+        printf("\tx=%.10e\n", x);
+        printf("\ty=%.10e\n", y);
+        printf("\tpx=%.10e\n", px);
+        printf("\tpy=%.10e\n", py);
+        printf("\tz=%.10e\n", zeta);
+        printf("\tpzeta=%.20e\n", pzeta);   
+        printf("\tdelta=%.20e\n", delta);   
+*/
         // Change reference frame
         change_ref_frame_coordinates(
             &x, &px, &y, &py, &zeta, &pzeta,
             shift_x, shift_px, shift_y, shift_py, shift_zeta, shift_pzeta,
             sin_phi, cos_phi, tan_phi, sin_alpha, cos_alpha);
 
+/*
+        printf("[beambeam3d] [%d] after boost:\n", part->ipart);
+        printf("\tx=%.10e\n", x);
+        printf("\ty=%.10e\n", y);
+        printf("\tpx=%.10e\n", px);
+        printf("\tpy=%.10e\n", py);
+        printf("\tz=%.10e\n", zeta);
+        printf("\tpzeta=%.20e\n", pzeta);   
+*/
+
+        // here pzeta is not updated but has to be! changes delta
+        LocalParticle_update_pzeta(part, pzeta);
```

Comment:
I am afraid this slows down simulations without beamstrahlung. It would be better to compute the energy veriables internally one from the other without updating the particle

Reply:
Restructured the code and removed the calls to pzeta. If there is no BS, there are no extra calls. If there is BS:

<pre>
// emit beamstrahlung photons from single macropart
if (flag_beamstrahlung==1){
LocalParticle_update_pzeta(part, *pzeta_star);  // update energy vars with boost and/or last kick
double const Fr = hypot(Fx_star, Fy_star) * LocalParticle_get_rpp(part); // total kick [1]
/*gpuglmem*/ double const dz = .5*BeamBeamBiGaussian3DData_get_slices_other_beam_zeta_bin_width_star(el, i_slice);  // bending radius [m]
beamstrahlung(part, beamstrahlung_record, beamstrahlung_table_index, beamstrahlung_table, Fr, dz);
*pzeta_star = LocalParticle_get_pzeta(part);  // BS rescales energy vars, so load again before kick 
}
</pre>